File: src/main/java/org/modelingvalue/nelumbo/syntax/Parser.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.syntax;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.KnowledgeBase;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.Type;
import org.modelingvalue.nelumbo.syntax.Tokenizer.TokenizerResult;

public final class Parser implements ParseExceptionHandler {

    public static List<Node> parse(String string) throws ParseException {
        Tokenizer tokenizer = new Tokenizer(string + "\n", string);
        return new Parser(tokenizer.tokenize()).parseEvaluate().roots();
    }

    public static List<Node> parse(Class<?> clss, String fileName) throws ParseException {
        try {
            InputStream stream = clss.getResourceAsStream(fileName);
            if (stream == null) {
                throw new ParseException("Nelumbo resource " + fileName + " does not exist", fileName);
            }
            InputStream buffer = new BufferedInputStream(stream);
            String base = new String(buffer.readAllBytes());
            return new Parser(new Tokenizer(base, fileName).tokenize()).parseEvaluate().roots();
        } catch (IOException e) {
            throw new ParseException(e, "IOException during parse", fileName);
        }
    }

    public static ParserResult parse(TokenizerResult tokenizerResult) {
        return parse(KnowledgeBase.BASE, tokenizerResult);
    }

    public static ParserResult parse(KnowledgeBase knowledgeBase, TokenizerResult tokenizerResult) {
        ParserResult[] pra = new ParserResult[1];
        knowledgeBase.run(() -> pra[0] = new Parser(tokenizerResult).parseNonThrowing());
        return pra[0];
    }

    // Instance

    private final KnowledgeBase   knowledgeBase;
    private final TokenizerResult tokenizerResult;

    private ParserResult          result;

    public Parser(TokenizerResult tokenizerResult) {
        this.knowledgeBase = KnowledgeBase.CURRENT.get();
        this.tokenizerResult = tokenizerResult;
    }

    public ParserResult parseNonThrowing() {
        try {
            return parse(new ParserResult(tokenizerResult, false));
        } catch (ParseException e) {
            throw new IllegalStateException(e);
        }
    }

    public ParserResult parseThrowing() throws ParseException {
        return parse(new ParserResult(tokenizerResult, true));
    }

    public ParserResult parseEvaluate() throws ParseException {
        ParserResult parserResult = parse(new ParserResult(tokenizerResult, true));
        parserResult.evaluate();
        return parserResult;
    }

    private ParserResult parse(ParserResult result) throws ParseException {
        this.result = result;
        knowledgeBase.setExceptionHandler(this);
        try {
            Token token = tokenizerResult.first();
            ParseContext ctx = ParseContext.of(Type.TOP_GROUP, Integer.MIN_VALUE, knowledgeBase.parseContext());
            Node node = parseNode(token, ctx, null);
            if (node != null) {
                result.setRoot(node);
                token = node.nextToken();
                if (token != null) {
                    addException(new ParseException("Unexpected token " + token + " after end of input", token));
                }
            } else if (exceptions().isEmpty()) {
                addException(new ParseException("No syntax pattern found for " + token, token));
            }
            result.checkAssertions();
            return result;
        } finally {
            knowledgeBase.endParsing();
            this.result = null;
            tokenizerResult.checkAssertions();
        }
    }

    protected Node parseNode(Token token, ParseContext inner, ParseContext outer) throws ParseException {
        PatternResult result = new PatternResult(this, inner);
        if (!preParse(token, null, result, outer)) {
            return null;
        }
        Node left = result.postParse();
        token = left != null && inner.precedence() < Integer.MAX_VALUE ? left.nextToken() : null;
        while (token != null && preParse(token, left, result, null)) {
            if (inner.precedence() >= result.leftPrecedence()) {
                return left;
            }
            left = result.postParse();
            token = left != null ? left.nextToken() : null;
        }
        return left;
    }

    private boolean preParse(Token token, Node left, PatternResult result, ParseContext outer) throws ParseException {
        String group = result.context().group();
        if (outer != null) {
            return outer.preParse(group, token, left, result);
        }
        for (ParseContext pc = result.context(); pc != null; pc = pc.outer()) {
            if (pc.preParse(group, token, left, result)) {
                return true;
            }
        }
        return false;
    }

    public KnowledgeBase knowledgeBase() {
        return knowledgeBase;
    }

    @Override
    public void addException(ParseException exception) throws ParseException {
        result.addException(exception);
    }

    @Override
    public List<ParseException> exceptions() {
        return result.exceptions();
    }

}

File: web-ts/src/syntax/Parser.ts
/**
 * Parser - precedence climbing parser.
 * @JAVA_REF org.modelingvalue.nelumbo.syntax.Parser
 */

import { List } from 'immutable';
import type { Token } from './Token';
import type { Node } from '../Node';
import { TOP_GROUP } from '../Type';
import { Variable } from '../Variable';
import { Tokenizer, type TokenizerResult } from './Tokenizer';
import type { KnowledgeBase } from '../KnowledgeBase';
import type { ParseContext } from './ParseContext';
import { createParseContext } from './ParseContext';
import type { ParseState } from './ParseState';
import { PatternResult } from './PatternResult';
import { ParserResult } from './ParserResult';
import { ParseException } from './ParseException';

/**
 * Parser - parses tokenized input using precedence climbing.
 */
export class Parser {
  private readonly _knowledgeBase: KnowledgeBase;
  private readonly _tokenizerResult: TokenizerResult;

  private _result: ParserResult | null = null;

  constructor(knowledgeBase: KnowledgeBase, tokenizerResult: TokenizerResult) {
    this._knowledgeBase = knowledgeBase;
    this._tokenizerResult = tokenizerResult;
  }

  /**
   * Parse without throwing on errors.
   */
  parseNonThrowing(): ParserResult {
    try {
      return this.parse(new ParserResult(this._tokenizerResult, false), false);
    } catch (e) {
      if (e instanceof ParseException) {
        throw new Error('Unexpected ParseException in non-throwing mode: ' + e.message);
      }
      throw e;
    }
  }

  /**
   * Parse and throw on errors.
   */
  parseThrowing(): ParserResult {
    return this.parse(new ParserResult(this._tokenizerResult, true), false);
  }

  /**
   * Parse and evaluate.
   */
  parseEvaluate(): ParserResult {
    const result = this.parse(new ParserResult(this._tokenizerResult, true), false);
    result.evaluate();
    return result;
  }

  /**
   * Parse multiple roots without throwing.
   */
  parseMultipleNonThrowing(): ParserResult {
    try {
      return this.parse(new ParserResult(this._tokenizerResult, false), true);
    } catch (e) {
      if (e instanceof ParseException) {
        throw new Error('Unexpected ParseException in non-throwing mode: ' + e.message);
      }
      throw e;
    }
  }

  private parse(result: ParserResult, multiple: boolean): ParserResult {
    return this._knowledgeBase.run(() => {
      this._result = result;
      this._knowledgeBase.setExceptionHandler(this);
      Parser._parseNodeCount = 0;

      try {
        let token = this._tokenizerResult.first;
        // @JAVA_REF Parser.parse() — use two-arg ParseContext.of(group, precedence)
        const ctx = createParseContext(TOP_GROUP, Number.MIN_SAFE_INTEGER);
        const node = this.parseNode(token, ctx);

        if (node !== null) {
          result.setRoot(node);
          token = node.nextToken();
          if (token !== null) {
            this.addException(
              ParseException.fromToken('Unexpected token ' + token + ' after end of input', token)
            );
          }
        } else if (this.exceptions().isEmpty()) {
          this.addException(
            ParseException.fromToken('No syntax pattern found for ' + token, token!)
          );
        }

        result.checkAssertions();
        return result;
      } finally {
        this._knowledgeBase.endParsing(multiple);
        this._result = null;
      }
    });
  }

  /**
   * Parse a single node.
   */
  private static _parseNodeDepth = 0;
  private static _parseNodeCount = 0;
  // @JAVA_REF Parser.parseNode(Token, ParseContext)
  parseNode(token: Token | null, ctx: ParseContext): Node | null {
    Parser._parseNodeDepth++;
    Parser._parseNodeCount++;

    // Safety limits to prevent browser hangs
    if (Parser._parseNodeDepth > 1000) {
      Parser._parseNodeDepth--;
      throw new Error('Parser.parseNode: Maximum recursion depth exceeded');
    }
    if (Parser._parseNodeCount > 50000) {
      throw new Error('Parser.parseNode: Maximum call count exceeded');
    }

    try {
      // Pre-parse phase
      let result = this.preParse(token, ctx, null);
      if (result === null) {
        return null;
      }

      // Post-parse phase
      let left = result.postParse(ctx);
      if (left !== null && ctx.precedence() < Number.MAX_SAFE_INTEGER) {
        token = left.nextToken();
        if (token !== null) {
          result = this.preParse(token, ctx, left);

          while (result !== null) {
            if (ctx.precedence() >= (result.leftPrecedence() ?? Number.MAX_SAFE_INTEGER)) {
              return left;
            }

            left = result.postParse(ctx);
            if (left === null) {
              break;
            }

            token = left.nextToken();
            if (token === null) {
              return left;
            }

            result = this.preParse(token, ctx, left);
          }
        }
      }

      return left;
    } finally {
      Parser._parseNodeDepth--;
    }
  }

  /**
   * Get group state from knowledge base.
   */
  groupState(group: string): ParseState | null {
    return this._knowledgeBase.groupState(group);
  }

  /**
   * Look up a variable.
   */
  variable(token: Token, ctx: ParseContext): Variable | null {
    return this._knowledgeBase.variable(token, ctx.group(), this);
  }

  /**
   * Pre-parse phase.
   */
  preParse(token: Token | null, ctx: ParseContext, left: Node | null): PatternResult | null {
    return this._knowledgeBase.preParse(token, ctx, left, this);
  }

  /**
   * Get the knowledge base.
   */
  knowledgeBase(): KnowledgeBase {
    return this._knowledgeBase;
  }

  /**
   * Add an exception.
   */
  addException(exception: ParseException): void {
    if (this._result !== null) {
      this._result.addException(exception);
    }
  }

  /**
   * Get all exceptions.
   */
  exceptions(): List<ParseException> {
    return this._result?.exceptions() ?? List();
  }
}

/**
 * Static parse methods.
 */
export function parseString(knowledgeBase: KnowledgeBase, input: string): List<Node> {
  const tokenizer = new Tokenizer(input + '\n', input);
  const parser = new Parser(knowledgeBase, tokenizer.tokenize());
  return parser.parseEvaluate().roots();
}

export function parseTokenizerResult(knowledgeBase: KnowledgeBase, tokenizerResult: TokenizerResult): ParserResult {
  return new Parser(knowledgeBase, tokenizerResult).parseNonThrowing();
}
